packages/mui-joy/src/TabPanel/TabPanel.tsx

```diff
@@ -84,7 +94,11 @@ const TabPanel = React.forwardRef(function TabPanel(inProps, ref) {
     className: classes.root,
   });
 
-  return <TabPanelRoot {...tabPanelRootProps}>{!hidden && children}</TabPanelRoot>;
+  if (hidden && !keepMounted) {
+    return null;
+  }
+
+  return <TabPanelRoot {...tabPanelRootProps}>{children}</TabPanelRoot>;
```

Comment:
We might need to wait until TabList uses React context (it is currently using `React.cloneElement`). cc @michaldudak

Reply:
Issue for reference: #35084